### PR TITLE
[TP] verify numeric parity on Transfromers for multiple iterations

### PR DESCRIPTION
**Pull request by weifengpy**
Before setting up float8 numeric parity test, I have to set up regular TP numeric parity test, preferrably testing 10 iterations

this PR sets a baseline of TP numerics. I can verify fp8 on top of it

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132543
* #132350

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: test/distributed/tensor/parallel/test_tp_examples.py [hunk 1]
```diff
@@ -215,75 +226,72 @@ def test_transformer_training(self, is_seq_parallel=False):
...
-                },
-            )
-        else:
-            self.assertDictEqual(
-                comm_mode.get_comm_counts(),
-                {
-                    c10d_functional.all_reduce: 6,
-                    c10d_functional.all_gather_into_tensor: 1,
-                },
+        steps = 10 if dtype is torch.float64 else 1
```
Comment (weifengpy) [inline test/distributed/tensor/parallel/test_tp_examples.py:229]: to verify numeric parity for multiple iterations (eg 10), I have to bring back `troch.float64` from previous PR

for float64, I also bring back `module_tp.model_args.n_heads // device_mesh.size()`
for float32, it's still DTensor as input for attention

File: test/distributed/tensor/parallel/test_tp_examples.py [hunk 2]
```diff
@@ -80,12 +80,16 @@ def _test_mlp_training_e2e(self, is_seq_parallel=False, recompute_activation=Fal
             torch.arange(0, NUM_DEVICES),
         )
         parallelize_plan = {
-            "net1": ColwiseParallel(input_layouts=Shard(0))
```
Comment (weifengpy) [inline test/distributed/tensor/parallel/test_tp_examples.py:83]: linter change

File: test/distributed/tensor/parallel/test_tp_examples.py (hunk 1 above)
Comment (tianyu-l) [reply test/distributed/tensor/parallel/test_tp_examples.py:229]: For float8, are you still going to use the math attention?
For fp16 IIRC there is numerical issues with that https://github.com/pytorch/torchtitan/issues/317

A related issue is #125741

Comment (weifengpy) [reply test/distributed/tensor/parallel/test_tp_examples.py:229]: the related issues are very helpful! didnt know there was stories behind this

for float8 model, sdpa is still in high precision ( thus can be math attention in fp64)

Comment (tianyu-l) [reply test/distributed/tensor/parallel/test_tp_examples.py:229]: I discussed the incompatibility between DTensor and math attention, and the conclusion is we should try modifying math attention code to unblock DTensor.

In the very short term, do we feel we have to test fp8 for math attention? If that's the case, I'm ok with adding back the "hacks" for now.

Comment (weifengpy) [reply test/distributed/tensor/parallel/test_tp_examples.py:229]: > we should try modifying math attention code to unblock DTensor

that sounds exciting

> In the very short term, do we feel we have to test fp8 for math attention? If that's the case, I'm ok with adding back the "hacks" for now.

Yes, I prefer to have a 10-iteration working now so I can build on top of it. Once DTensor + math attention worked, we can remove the hack around `n_heads // device_mesh.size()`. what do you think ?

**Review by tianyu-l — APPROVED** [review]
sgtm

**Pull request closed without merging**